server/src/main/java/org/opensearch/search/aggregations/BucketCollectorProcessor.java

```diff
@@ -72,6 +74,20 @@ public void processPostCollection(Collector collectorTree) throws IOException {
                 }
             } else if (currentCollector instanceof BucketCollector) {
                 ((BucketCollector) currentCollector).postCollection();
+
+                // Perform build aggregation during post collection
+                if (currentCollector instanceof AggregatorBase) {
```

Comment:
In the same vein, AFAICT the `setInternalAggregation` is always build from `buildTopLevel`, why do we need this method (and associated state) at all?

Reply:
Thanks @reta!
At a high level I want to create the `InternalAggregation` object on the `index_searcher` thread during `processPostCollection` and then I need to retrieve said object on the `search` thread during `reduce`. Because `CollectorManager::reduce` only takes a list of collectors as input, we need to temporarily store this state in the collector (`Aggregator`) itself.
As for your question about why not do this in `Collector::postCollection` itself, the problem is that `postCollection` is implemented per-collector so for example the `ParentJoinAggregator` overrides the default implementation.
I do believe you are right that `AggregatorBase::buildAndSetInternalAggregation` itself is not necessary and we only need the methods from `Aggregator`, I think that's just an artifact from when I was trying to implement this in `Collector::postCollection`.